crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -1373,42 +1305,161 @@ pub trait SemanticSyntaxContext {
     /// Return the [`TextRange`] at which a name is declared as `global` in the current scope.
     fn global(&self, name: &str) -> Option<TextRange>;
 
+    /// Returns `true` if the visitor is currently in an async context, i.e. an async function.
+    fn in_async_context(&self) -> bool;
+
+    /// Returns `true` if the visitor is currently inside of a synchronous comprehension.
+    ///
+    /// This method is necessary because `in_async_context` only checks for the nearest, enclosing
+    /// function to determine the (a)sync context. Instead, this method will search all enclosing
+    /// scopes until it finds a sync comprehension. As a result, the two methods will typically be
+    /// used together.
+    fn in_sync_comprehension(&self) -> bool;
+
+    /// Returns `true` if the visitor is at the top-level module scope.
+    fn in_module_scope(&self) -> bool;
+
+    /// Returns `true` if the source file is a Jupyter notebook.
+    fn in_notebook(&self) -> bool;
+
     fn report_semantic_error(&self, error: SemanticSyntaxError);
 }
 
+enum Scope {
+    Module,
+    Function { is_async: bool },
+    Comprehension { is_async: bool },
+}
+
 #[derive(Default)]
-pub struct SemanticSyntaxCheckerVisitor<Ctx> {
+pub struct SemanticSyntaxCheckerVisitor<'a> {
```

Comment:
Should we add a docblock stating that this visitor is only for testing?

Or should we move it to the tests module, considering that it isn't used anywhere else?